File: ExamenP1/01.SOAP_JAVA/01.SERVIDOR/FederacionFutbol_WS/src/main/java/ec/edu/monster/servicio/FederacionService.java
package ec.edu.monster.servicio;

import ec.edu.monster.modelo.*;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;

/**
 * Capa de servicio/DAO para acceso a la base de datos MySQL.
 * Maneja todas las operaciones CRUD de la Federación de Fútbol.
 */
public class FederacionService {

    private static final String URL = "jdbc:mysql://localhost:3306/ticketpremium_db?useSSL=false&serverTimezone=America/Guayaquil&allowPublicKeyRetrieval=true";
    private static final String USER = "root";
    private static final String PASS = "root";

    static {
        try {
            Class.forName("com.mysql.cj.jdbc.Driver");
        } catch (ClassNotFoundException e) {
            System.err.println("ERROR: Driver MySQL no encontrado: " + e.getMessage());
        }
    }

    private Connection getConnection() throws SQLException {
        return DriverManager.getConnection(URL, USER, PASS);
    }

    // ==========================================
    // PARTIDOS DISPONIBLES (fecha >= ahora)
    // ==========================================
    public List<PartidoFutbol> obtenerPartidosDisponibles() {
        List<PartidoFutbol> partidos = new ArrayList<>();
        String sql = "SELECT CODIGO, EQUIPO_LOCAL, EQUIPO_VISITA, "
                   + "DATE_FORMAT(FECHA, '%Y-%m-%d %H:%i') AS FECHA, LUGAR "
                   + "FROM PARTIDO_FUTBOL WHERE FECHA >= NOW() ORDER BY FECHA";
        try (Connection conn = getConnection();
             PreparedStatement ps = conn.prepareStatement(sql);
             ResultSet rs = ps.executeQuery()) {
            while (rs.next()) {
                PartidoFutbol p = new PartidoFutbol();
                p.setCodigo(rs.getInt("CODIGO"));
                p.setEquipoLocal(rs.getString("EQUIPO_LOCAL"));
                p.setEquipoVisita(rs.getString("EQUIPO_VISITA"));
                p.setFecha(rs.getString("FECHA"));
                p.setLugar(rs.getString("LUGAR"));
                partidos.add(p);
            }
        } catch (SQLException e) {
            System.err.println("Error obtenerPartidosDisponibles: " + e.getMessage());
        }
        return partidos;
    }

    // ==========================================
    // LOCALIDADES DISPONIBLES PARA UN PARTIDO
    // ==========================================
    public List<LocalidadPartido> obtenerLocalidades(int codigoPartido) {
        List<LocalidadPartido> localidades = new ArrayList<>();
        String sql = "SELECT ID, CODIGO_PARTIDO, CODIGO_LOCALIDAD, DISPONIBILIDAD, PRECIO "
                   + "FROM LOCALIDAD_PARTIDO "
                   + "WHERE CODIGO_PARTIDO = ? AND DISPONIBILIDAD > 0";
        try (Connection conn = getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            ps.setInt(1, codigoPartido);
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    LocalidadPartido l = new LocalidadPartido();
                    l.setId(rs.getInt("ID"));
                    l.setCodigoPartido(rs.getInt("CODIGO_PARTIDO"));
                    l.setCodigoLocalidad(rs.getString("CODIGO_LOCALIDAD"));
                    l.setDisponibilidad(rs.getInt("DISPONIBILIDAD"));
                    l.setPrecio(rs.getDouble("PRECIO"));
                    localidades.add(l);
                }
            }
        } catch (SQLException e) {
            System.err.println("Error obtenerLocalidades: " + e.getMessage());
        }
        return localidades;
    }

    // ==========================================
    // DECREMENTAR DISPONIBILIDAD
    // ==========================================
    public boolean decrementarDisponibilidad(int idLocalidad, int cantidad) {
        String sql = "UPDATE LOCALIDAD_PARTIDO "
                   + "SET DISPONIBILIDAD = DISPONIBILIDAD - ? "
                   + "WHERE ID = ? AND DISPONIBILIDAD >= ?";
        try (Connection conn = getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            ps.setInt(1, cantidad);
            ps.setInt(2, idLocalidad);
            ps.setInt(3, cantidad);
            return ps.executeUpdate() > 0;
        } catch (SQLException e) {
            System.err.println("Error decrementarDisponibilidad: " + e.getMessage());
            return false;
        }
    }

    // ==========================================
    // REGISTRAR COMPRA (Factura + Detalles)
    // ==========================================
    public Factura registrarCompra(int codigoPartido, String nombreCliente,
                                   String codigoLocalidad, int cantidad, double precioUnitario) {
        double subtotal = cantidad * precioUnitario;
        double iva = subtotal * 0.15;          // IVA Ecuador 15%
        double total = subtotal + iva;

        Factura factura = new Factura();
        factura.setCodigoPartido(codigoPartido);
        factura.setNombreCliente(nombreCliente);
        factura.setSubtotal(subtotal);
        factura.setIva(iva);
        factura.setTotal(total);

        String sqlFactura = "INSERT INTO FACTURA (CODIGO_PARTIDO, NOMBRE_CLIENTE, FECHA, SUBTOTAL, IVA, TOTAL) "
                          + "VALUES (?, ?, NOW(), ?, ?, ?)";
        String sqlDetalle = "INSERT INTO DETALLE_FACTURA (ID_FACTURA, CODIGO_PARTIDO, CODIGO_LOCALIDAD, CANTIDAD, PRECIO_UNITARIO, SUBTOTAL) "
                          + "VALUES (?, ?, ?, ?, ?, ?)";

        try (Connection conn = getConnection()) {
            conn.setAutoCommit(false);
            try {
                // Insertar factura
                try (PreparedStatement ps = conn.prepareStatement(sqlFactura, Statement.RETURN_GENERATED_KEYS)) {
                    ps.setInt(1, codigoPartido);
                    ps.setString(2, nombreCliente);
                    ps.setDouble(3, subtotal);
                    ps.setDouble(4, iva);
                    ps.setDouble(5, total);
                    ps.executeUpdate();

                    try (ResultSet rs = ps.getGeneratedKeys()) {
                        if (rs.next()) {
                            factura.setId(rs.getInt(1));
                        }
                    }
                }

                // Insertar detalle
                try (PreparedStatement ps = conn.prepareStatement(sqlDetalle)) {
                    ps.setInt(1, factura.getId());
                    ps.setInt(2, codigoPartido);
                    ps.setString(3, codigoLocalidad);
                    ps.setInt(4, cantidad);
                    ps.setDouble(5, precioUnitario);
                    ps.setDouble(6, subtotal);
                    ps.executeUpdate();
                }

                conn.commit();
            } catch (SQLException e) {
                conn.rollback();
                System.err.println("Error registrarCompra (rollback): " + e.getMessage());
                return null;
            }
        } catch (SQLException e) {
            System.err.println("Error registrarCompra: " + e.getMessage());
            return null;
        }
        return factura;
    }

    // ==========================================
    // OBTENER PARTIDO POR CÓDIGO
    // ==========================================
    public PartidoFutbol obtenerPartido(int codigoPartido) {
        String sql = "SELECT CODIGO, EQUIPO_LOCAL, EQUIPO_VISITA, "
                   + "DATE_FORMAT(FECHA, '%Y-%m-%d %H:%i') AS FECHA, LUGAR "
                   + "FROM PARTIDO_FUTBOL WHERE CODIGO = ?";
        try (Connection conn = getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            ps.setInt(1, codigoPartido);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    PartidoFutbol p = new PartidoFutbol();
                    p.setCodigo(rs.getInt("CODIGO"));
                    p.setEquipoLocal(rs.getString("EQUIPO_LOCAL"));
                    p.setEquipoVisita(rs.getString("EQUIPO_VISITA"));
                    p.setFecha(rs.getString("FECHA"));
                    p.setLugar(rs.getString("LUGAR"));
                    return p;
                }
            }
        } catch (SQLException e) {
            System.err.println("Error obtenerPartido: " + e.getMessage());
        }
        return null;
    }

    // ==========================================
    // RESUMEN DE VENTAS DE UN PARTIDO
    // ==========================================
    public List<ResumenVenta> obtenerResumenVentas(int codigoPartido) {
        List<ResumenVenta> resumen = new ArrayList<>();
        String sql = "SELECT CODIGO_LOCALIDAD, "
                   + "SUM(CANTIDAD) AS VENDIDOS, "
                   + "SUM(SUBTOTAL) AS TOTAL_RECAUDADO "
                   + "FROM DETALLE_FACTURA "
                   + "WHERE CODIGO_PARTIDO = ? "
                   + "GROUP BY CODIGO_LOCALIDAD";
        try (Connection conn = getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            ps.setInt(1, codigoPartido);
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    ResumenVenta r = new ResumenVenta();
                    r.setCodigoLocalidad(rs.getString("CODIGO_LOCALIDAD"));
                    r.setVendidos(rs.getInt("VENDIDOS"));
                    r.setTotalRecaudado(rs.getDouble("TOTAL_RECAUDADO"));
                    resumen.add(r);
                }
            }
        } catch (SQLException e) {
            System.err.println("Error obtenerResumenVentas: " + e.getMessage());
        }
        return resumen;
    }
}

File: ExamenP1/02.RESTFUL_DOTNET/01.SERVIDOR/FederacionFutbol_API/Servicio/FederacionService.cs
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Data.SqlClient;
using FederacionFutbol_API.Models;

namespace FederacionFutbol_API.Servicio
{
    /// <summary>
    /// Capa de servicio/DAO para acceso a SQL Server.
    /// </summary>
    public class FederacionService
    {
        private static readonly string ConnStr =
            ConfigurationManager.ConnectionStrings["TicketPremiumDB"].ConnectionString;

        // ==========================================
        // PARTIDOS DISPONIBLES
        // ==========================================
        public List<PartidoFutbol> ObtenerPartidosDisponibles()
        {
            var lista = new List<PartidoFutbol>();
            using (var conn = new SqlConnection(ConnStr))
            {
                conn.Open();
                var cmd = new SqlCommand(
                    "SELECT CODIGO, EQUIPO_LOCAL, EQUIPO_VISITA, " +
                    "FORMAT(FECHA, 'yyyy-MM-dd HH:mm') AS FECHA, LUGAR " +
                    "FROM PARTIDO_FUTBOL WHERE FECHA >= GETDATE() ORDER BY FECHA", conn);
                using (var rd = cmd.ExecuteReader())
                {
                    while (rd.Read())
                    {
                        lista.Add(new PartidoFutbol
                        {
                            Codigo = rd.GetInt32(0),
                            EquipoLocal = rd.GetString(1),
                            EquipoVisita = rd.GetString(2),
                            Fecha = rd.GetString(3),
                            Lugar = rd.GetString(4)
                        });
                    }
                }
            }
            return lista;
        }

        // ==========================================
        // LOCALIDADES DISPONIBLES
        // ==========================================
        public List<LocalidadPartido> ObtenerLocalidades(int codigoPartido)
        {
            var lista = new List<LocalidadPartido>();
            using (var conn = new SqlConnection(ConnStr))
            {
                conn.Open();
                var cmd = new SqlCommand(
                    "SELECT ID, CODIGO_PARTIDO, CODIGO_LOCALIDAD, DISPONIBILIDAD, PRECIO " +
                    "FROM LOCALIDAD_PARTIDO WHERE CODIGO_PARTIDO = @cod AND DISPONIBILIDAD > 0", conn);
                cmd.Parameters.AddWithValue("@cod", codigoPartido);
                using (var rd = cmd.ExecuteReader())
                {
                    while (rd.Read())
                    {
                        lista.Add(new LocalidadPartido
                        {
                            Id = rd.GetInt32(0),
                            CodigoPartido = rd.GetInt32(1),
                            CodigoLocalidad = rd.GetString(2),
                            Disponibilidad = rd.GetInt32(3),
                            Precio = rd.GetDecimal(4)
                        });
                    }
                }
            }
            return lista;
        }

        // ==========================================
        // DECREMENTAR DISPONIBILIDAD
        // ==========================================
        public bool DecrementarDisponibilidad(int idLocalidad, int cantidad)
        {
            using (var conn = new SqlConnection(ConnStr))
            {
                conn.Open();
                var cmd = new SqlCommand(
                    "UPDATE LOCALIDAD_PARTIDO SET DISPONIBILIDAD = DISPONIBILIDAD - @cant " +
                    "WHERE ID = @id AND DISPONIBILIDAD >= @cant", conn);
                cmd.Parameters.AddWithValue("@cant", cantidad);
                cmd.Parameters.AddWithValue("@id", idLocalidad);
                return cmd.ExecuteNonQuery() > 0;
            }
        }

        // ==========================================
        // REGISTRAR COMPRA
        // ==========================================
        public Factura RegistrarCompra(CompraRequest req)
        {
            decimal subtotal = req.Cantidad * req.PrecioUnitario;
            decimal iva = subtotal * 0.15m;
            decimal total = subtotal + iva;

            using (var conn = new SqlConnection(ConnStr))
            {
                conn.Open();
                var tran = conn.BeginTransaction();
                try
                {
                    // Decrementar
                    var cmdDec = new SqlCommand(
                        "UPDATE LOCALIDAD_PARTIDO SET DISPONIBILIDAD = DISPONIBILIDAD - @cant " +
                        "WHERE ID = @id AND DISPONIBILIDAD >= @cant", conn, tran);
                    cmdDec.Parameters.AddWithValue("@cant", req.Cantidad);
                    cmdDec.Parameters.AddWithValue("@id", req.IdLocalidad);
                    if (cmdDec.ExecuteNonQuery() <= 0) { tran.Rollback(); return null; }

                    // Insertar Factura
                    var cmdFact = new SqlCommand(
                        "INSERT INTO FACTURA (CODIGO_PARTIDO, NOMBRE_CLIENTE, FECHA, SUBTOTAL, IVA, TOTAL) " +
                        "VALUES (@cod, @nom, GETDATE(), @sub, @iva, @tot); SELECT SCOPE_IDENTITY();", conn, tran);
                    cmdFact.Parameters.AddWithValue("@cod", req.CodigoPartido);
                    cmdFact.Parameters.AddWithValue("@nom", req.NombreCliente);
                    cmdFact.Parameters.AddWithValue("@sub", subtotal);
                    cmdFact.Parameters.AddWithValue("@iva", iva);
                    cmdFact.Parameters.AddWithValue("@tot", total);
                    int idFactura = Convert.ToInt32(cmdFact.ExecuteScalar());

                    // Insertar Detalle
                    var cmdDet = new SqlCommand(
                        "INSERT INTO DETALLE_FACTURA (ID_FACTURA, CODIGO_PARTIDO, CODIGO_LOCALIDAD, CANTIDAD, PRECIO_UNITARIO, SUBTOTAL) " +
                        "VALUES (@idf, @cod, @loc, @cant, @pu, @sub)", conn, tran);
                    cmdDet.Parameters.AddWithValue("@idf", idFactura);
                    cmdDet.Parameters.AddWithValue("@cod", req.CodigoPartido);
                    cmdDet.Parameters.AddWithValue("@loc", req.CodigoLocalidad);
                    cmdDet.Parameters.AddWithValue("@cant", req.Cantidad);
                    cmdDet.Parameters.AddWithValue("@pu", req.PrecioUnitario);
                    cmdDet.Parameters.AddWithValue("@sub", subtotal);
                    cmdDet.ExecuteNonQuery();

                    tran.Commit();
                    return new Factura
                    {
                        Id = idFactura,
                        CodigoPartido = req.CodigoPartido,
                        NombreCliente = req.NombreCliente,
                        Subtotal = subtotal,
                        Iva = iva,
                        Total = total
                    };
                }
                catch
                {
                    tran.Rollback();
                    return null;
                }
            }
        }

        // ==========================================
        // OBTENER PARTIDO
        // ==========================================
        public PartidoFutbol ObtenerPartido(int codigo)
        {
            using (var conn = new SqlConnection(ConnStr))
            {
                conn.Open();
                var cmd = new SqlCommand(
                    "SELECT CODIGO, EQUIPO_LOCAL, EQUIPO_VISITA, " +
                    "FORMAT(FECHA, 'yyyy-MM-dd HH:mm') AS FECHA, LUGAR " +
                    "FROM PARTIDO_FUTBOL WHERE CODIGO = @cod", conn);
                cmd.Parameters.AddWithValue("@cod", codigo);
                using (var rd = cmd.ExecuteReader())
                {
                    if (rd.Read())
                    {
                        return new PartidoFutbol
                        {
                            Codigo = rd.GetInt32(0),
                            EquipoLocal = rd.GetString(1),
                            EquipoVisita = rd.GetString(2),
                            Fecha = rd.GetString(3),
                            Lugar = rd.GetString(4)
                        };
                    }
                }
            }
            return null;
        }

        // ==========================================
        // RESUMEN DE VENTAS
        // ==========================================
        public List<ResumenVenta> ObtenerResumenVentas(int codigoPartido)
        {
            var lista = new List<ResumenVenta>();
            using (var conn = new SqlConnection(ConnStr))
            {
                conn.Open();
                var cmd = new SqlCommand(
                    "SELECT CODIGO_LOCALIDAD, SUM(CANTIDAD) AS VENDIDOS, " +
                    "SUM(SUBTOTAL) AS TOTAL_RECAUDADO " +
                    "FROM DETALLE_FACTURA WHERE CODIGO_PARTIDO = @cod " +
                    "GROUP BY CODIGO_LOCALIDAD", conn);
                cmd.Parameters.AddWithValue("@cod", codigoPartido);
                using (var rd = cmd.ExecuteReader())
                {
                    while (rd.Read())
                    {
                        lista.Add(new ResumenVenta
                        {
                            CodigoLocalidad = rd.GetString(0),
                            Vendidos = rd.GetInt32(1),
                            TotalRecaudado = rd.GetDecimal(2)
                        });
                    }
                }
            }
            return lista;
        }
    }
}
